### DISCUSSION TITLE
Get $data from current component only, not all parent ones.

### DISCUSSION THREAD

**Discussion by yabdab** [alpinejs/alpine#3925] [1. Help]
So I have a parent component that wraps several child components.

The parent component has an array named "childArray".

Each child component in parent will load it's data  into the childArray from parent. on x-inut.

```
x-init="childArray.push($data)"
```

This works, but creates a circular structure since $data  contains the parent component data  AS WELL AS the child's.

Hope this make sense so far 😉

I tried doing the following to get only the child data into the array, but it  is undefined value ...

```
x-init="childArray.push($data[0])"
```

So my question is, how can I pass the the current , child component ,  data only into the childArray?

Thanks.

**Comment by bennadel** [comment]
I just ran into something similar. I was trying to create an `x-ref-append="{{ array }}"` directive that would append the current `element` to the given `array` reference. This creates a circular reference if there is an intermediary `x-data` because the "stack" of scopes start referencing each other. This doesn't always cause a problem; but, if anything has to be `$watch()`ed, there is an internal call to `JSON.stringify()` which forces the circular reference to throw an error.

**Comment by ekwoka** [comment]
I'm not sure what the goals intent would be here. This seems like XY problems.

Elaborating more on what the real goal is would be helpful.

How does the child get data in this case that it needs to add it's entire context to the parent?

How do you discern what is a child property vs a parent property?

Can you no do like x-for does and scope the child data under a key itself?

**Reply by bennadel** [reply]
@ekwoka I am just playing around with different mindsets. On the one hand, I can give Alpine some data, and then have it render a list (using `x-for`). Or, I can render the list of the server, and then let Alpine augment the list behavior. In the latter case (server-rendered list), I was going to see if I could use something like `x-ref-append` to aggregate the `<li>` elements into an array.

Mostly, at this point, I'm still just building up a mental model of what Alpine can be used to do. And, I'm trying to keep an eye on "progressive enhancement" (hence the server-rendered list). Just playing around.

**Reply by ekwoka** [reply]
I'd just have them add their keyed data to an array in the parent.

Like a reverse x-for

**Reply by bennadel** [reply]
Yeah, that's makes sense. So basically sending the data down twice (once in JSON form, once in HTML form) essentially.

I was also playing around with some ideas of constructing an object from the DOM on the fly using an `x-dom-object` directive:

```html
<tr x-dom-object="friend">
	<td x-dom-object="friend.id">
		#encodeForHtml( friend.id )#
	</td>
	<td x-dom-object="friend.name">
		#encodeForHtml( friend.name )#
	</td>
</tr>
```

The idea being that this would _somehow_ populate the `friend: { id: ..., name: ... }` using the DOM. I couldn't make an API that felt right. But, like I said, I'm just playing around with the framework to see what is possible.

**Comment by yabdab** [comment]
Here is what I ended up doing and it works pretty well so far...

In Markup ...

```html
<div x-data="item" x-init="addItemToArray($el)"> ... </div>
```

In App ...

```javascript
addItemToArray(el) {
      let item = Alpine.closestDataStack(el)[0];
      this.itemArray.push(item);
}
```
